src/jit/emitxarch.cpp

```diff
@@ -5170,158 +5208,191 @@ void emitter::emitIns_AX_R(instruction ins, emitAttr attr, regNumber ireg, regNu
+        emitIns_R_R_AR(ins, attr, reg, reg1, base, 0);
     }
     else
     {
         if (reg1 != reg)
         {
-            emitIns_R_R(INS_movaps, emitTypeSize(simdtype), reg, reg1);
+            emitIns_R_R(INS_movaps, attr, reg, reg1);
         }
-        emitIns_R_A(ins, emitTypeSize(simdtype), reg, indir, IF_RWR_ARD);
+        emitIns_R_AR(ins, attr, reg, base, 0);
     }
 }
 
-void emitter::emitIns_SIMD_R_R_AR(instruction ins, regNumber reg, regNumber reg1, regNumber base, var_types simdtype)
+void emitter::emitIns_SIMD_R_R_C(
+    instruction ins, emitAttr attr, regNumber reg, regNumber reg1, CORINFO_FIELD_HANDLE fldHnd, int offs)
 {
     if (UseVEXEncoding())
     {
-        emitIns_R_R_AR(ins, emitTypeSize(simdtype), reg, reg1, base, 0);
+        emitIns_R_R_C(ins, attr, reg, reg1, fldHnd, offs);
     }
     else
     {
         if (reg1 != reg)
         {
-            emitIns_R_R(INS_movaps, emitTypeSize(simdtype), reg, reg1);
+            emitIns_R_R(INS_movaps, attr, reg, reg1);
         }
-        emitIns_R_AR(ins, emitTypeSize(simdtype), reg, base, 0);
+        emitIns_R_C(ins, attr, reg, fldHnd, offs);
     }
 }
 
-void emitter::emitIns_SIMD_R_R_C(
-    instruction ins, regNumber reg, regNumber reg1, CORINFO_FIELD_HANDLE fldHnd, int offs, var_types simdtype)
+void emitter::emitIns_SIMD_R_R_R(instruction ins, emitAttr attr, regNumber reg, regNumber reg1, regNumber reg2)
 {
     if (UseVEXEncoding())
     {
-        emitIns_R_R_C(ins, emitTypeSize(simdtype), reg, reg1, fldHnd, offs);
+        emitIns_R_R_R(ins, attr, reg, reg1, reg2);
     }
     else
     {
         if (reg1 != reg)
         {
-            emitIns_R_R(INS_movaps, emitTypeSize(simdtype), reg, reg1);
+            emitIns_R_R(INS_movaps, attr, reg, reg1);
         }
-        emitIns_R_C(ins, emitTypeSize(simdtype), reg, fldHnd, offs);
+        emitIns_R_R(ins, attr, reg, reg2);
     }
 }
 
-void emitter::emitIns_SIMD_R_R_R(instruction ins, regNumber reg, regNumber reg1, regNumber reg2, var_types simdtype)
+void emitter::emitIns_SIMD_R_R_R_R(
+    instruction ins, emitAttr attr, regNumber reg, regNumber reg1, regNumber reg2, regNumber reg3)
 {
+    assert(isAvxBlendv(ins) || isSse41Blendv(ins));
     if (UseVEXEncoding())
     {
-        emitIns_R_R_R(ins, emitTypeSize(simdtype), reg, reg1, reg2);
+        // convert SSE encoding of SSE4.1 instructions to VEX encoding
+        switch (ins)
+        {
+            case INS_blendvps:
+                ins = INS_vblendvps;
+                break;
+            case INS_blendvpd:
+                ins = INS_vblendvpd;
+                break;
+            case INS_pblendvb:
+                ins = INS_vpblendvb;
+                break;
+            default:
+                break;
+        }
+        emitIns_R_R_R_R(ins, attr, reg, reg1, reg2, reg3);
     }
     else
     {
+        assert(isSse41Blendv(ins));
+        // SSE4.1 blendv* hardcode the mask vector (op3) in XMM0
+        if (reg3 != REG_XMM0)
```

Comment:
Do we have a bug tracking this? We should probably have a `TODO-XArch-CQ` with a link to it.

Reply:
As I understand based on above discussion and practice, this is not a bug, the current code can assign XMM0 to reg3 for simple cases. But in some complex use cases, when reg3 cannot get XMM0, LSRA won't insert the "copy" and codgen is responsible to do it.
BTW, the current CQ looks good to me, a few more zero-latency `movaps xmm, xmm` does not impact the performance too much. But there may be some register pressure issues that we need more workloads and tests to investigate.

Reply:
Yes, it is by convention that when a specific register is required, the code generator is responsible for checking whether that is what was assigned, and otherwise doing the copy. It has been debated whether all these checks are more or less efficient than if the register allocator inserted copies. But for now, this is how it's done. It's usually done in codegen (rather than the emitter), I believe, but for this case this seems like the right place.